scio-elasticsearch/src/main/java/com/spotify/scio/elasticsearch/ElasticsearchIO.java

```diff
@@ -0,0 +1,228 @@
+    /**
+     * Returns a tranform for writing to Elasticsearch cluster for a given name.
+     *
+     * @param clusterName name of the Elasticsearch cluster
+     */
+    public static Bound withClusterName(String clusterName) {
+      return new Bound().withClusterName(clusterName);
+    }
+
+    /**
+     * Returns a transform for writing to the Elasticsearch cluster for a given servers.
+     *
+     * @param servers endpoints for the Elasticsearch cluster
+     */
+    public static Bound withServers(InetSocketAddress[] servers) {
+      return new Bound().withServers(servers);
+    }
+
+    /**
+     * Returns a transform for writing to Elasticsearch cluster by providing
+     * slight delay specified by flushInterval.
+     *
+     * @param flushInterval delay applied to buffer elements. Defaulted to 1 seconds.
+     */
+    public static Bound withFlushInterval(Duration flushInterval) {
+      return new Bound().withFlushInterval(flushInterval);
+    }
+
+    public static class Bound extends PTransform<PCollection<IndexRequestWrapper>, PDone> {
+      private final String clusterName;
+      private final InetSocketAddress[] servers;
+      private final Duration flushInterval;
+
+      private Bound(final String clusterName,
+                    final InetSocketAddress[] servers,
+                    final Duration flushInterval) {
+        this.clusterName = clusterName;
+        this.servers = servers;
+        this.flushInterval = flushInterval == null? Duration.ofSeconds(1L): flushInterval;
+      }
+
+      Bound() {
+        this(null, null, null);
+      }
+
+      public Bound withClusterName(String clusterName) {
+        return new Bound(clusterName, servers, flushInterval);
+      }
+
+      public Bound withServers(InetSocketAddress[] servers) {
+        return new Bound(clusterName, servers, flushInterval);
+      }
+
+      public Bound withFlushInterval(Duration flushInterval) {
+        return new Bound(clusterName, servers, flushInterval);
+      }
+
+      @Override
+      public PDone expand(final PCollection<IndexRequestWrapper> input) {
+
+        if (clusterName == null) {
+          throw new IllegalStateException(
+              "need to set clustername of ElasticsearchIO.Write transform");
+        }
+
+        if (servers == null) {
+          throw new IllegalStateException(
+              "need to set clustername of ElasticsearchIO.Write transform");
+        }
+
+        input
+            .apply("Assign To Shard", ParDo.of(new AssignToShard()))
+            .apply("Re-Window to Global Window", Window.<KV<Long, IndexRequestWrapper>>into(new GlobalWindows())
+                       .triggering(Repeatedly.forever(
+                           AfterProcessingTime
+                               .pastFirstElementInPane()
+                               .plusDelayOf(javaToJoda(flushInterval))))
+                       .discardingFiredPanes())
+            .apply(GroupByKey.create())
```

Comment:
@lmayumi no need to implement such interface coz already exists as `ActionRequest` part of `elasticsearch` api. Every `IndexRequest` `DeleteRequest` `UpdateRequest` etc extends it. BulkRequests allow `[ActionRequest]`. You mention grouping by shard or type ... those should be left to user implementation ... and depending on the use case this `"sharding"` might not work ... so imo offering it as optional is better. 

My use case requires the above, and i think I'm failing to explain/convey my idea so i'll gladly purpose a PR after this one.